Add an output-base-name option to the Sphinx plot directive

asmeurer opened the pull request:
This allows specifying the output base name of the generated image files. The name can include '{counter}', which is automatically string formatted to an incrementing counter. The default if it is not specified is left intact as the current behavior, which is to use the base name of the provided script or the RST document.

This is required to use the plot directive with MyST, because the directive is broken with MyST (an issue I don't want to fix), requiring the use of eval-rst. But the way eval-rst works, the incrementing counter is not maintained across different eval-rst directives, meaning if you try to include multiple of them in the same document, the images will overwrite each other. This allows you to manually work around this with something like

````
```{eval-rst}
.. plot::
   :output-base-name: plot-1

   ...
```

```{eval-rst}
.. plot::
   :output-base-name: plot-2

   ...
```
````

Aside from this, it's generally useful to be able to specify the image name used for a plot, as a more informative name can be used rather than just `<document-name>-1.png`.

Note: I've done minimal testing of this, but more extensive testing would probably be a good idea. 

<!--
Thank you so much for your PR!  To help us review your contribution, please check
out the development guide https://matplotlib.org/devdocs/devel/index.html
-->

## PR summary
<!-- Please provide at least 1-2 sentences describing the pull request in detail
(Why is this change required?  What problem does it solve?) and link to relevant
issues and PRs.

Also please summarize the changes in the title, for example "Raise ValueError on
non-numeric input to set_xlim" and avoid non-descriptive titles such as "Addresses
issue #8576".
-->


## PR checklist
<!-- Please mark any checkboxes that do not apply to this PR as [N/A].-->

- [ ] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [ ] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [ ] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [ ] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [ ] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

<!--We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.-->

asmeurer commented on lib/matplotlib/sphinxext/plot_directive.py:
```diff
@@ -47,6 +47,15 @@
+    format ``{counter}`` to use an incremented counter. For example,
+    ``'plot-{counter}'`` will create files like ``plot-1.png``, ``plot-2.png``,
+    and so on. If the ``{counter}`` is not provided, two plots with the same
+    output-base-name may overwrite each other.
```
Perhaps the extension should actually check for this and just error.

asmeurer replied:
OK, nevermind, I undid this change. This seems harder to do than I thought, because you'd need to keep track of which files are generated.

asmeurer commented:
For anyone interested, the error you get if you try to use the plot directive with MyST directly is `ERROR: MockStateMachine has not yet implemented attribute 'insert_input'`.

asmeurer commented:
Ah, I didn't realize there actually already are unit tests for this. Let me update that.

asmeurer commented:
For me, the Sphinx extension tests fail on main

```
        writing output... [ 20%] included_plot_21
E
E       stderr:
E
E       Exception occurred:
E         File "/Users/aaronmeurer/miniconda3/envs/ndindex-matplotlib-dev/lib/python3.12/shutil.py", line 260, in copyfile
E           with open(src, 'rb') as fsrc:
E                ^^^^^^^^^^^^^^^
E       FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/wc/dppcpmxs1tlb36nqcw853wkm0000gn/T/pytest-of-aaronmeurer/pytest-2/test_srcset_version0/plot_directive/included_plot_21-1.2x.png'
E       The full traceback has been saved in /var/folders/wc/dppcpmxs1tlb36nqcw853wkm0000gn/T/sphinx-err-nut6ujx8.log, if you want to report the issue to the developers.
E       Please also report this if it was a user error, so that a better error message can be provided next time.
E       A bug report can be filed in the tracker at <https://github.com/sphinx-doc/sphinx/issues>. Thanks!
```

```
# Platform:         darwin; (macOS-14.4.1-arm64-arm-64bit)
# Sphinx version:   7.2.6
# Python version:   3.12.3 (CPython)
# Docutils version: 0.20.1
# Jinja2 version:   3.1.4
# Pygments version: 2.18.0

# Last messages:
#   
#   copying static files...
#   done
#   copying extra files...
#   done
#   done
#   [2K
#   writing output... [ 20%]
#   included_plot_21
#   

# Loaded extensions:
#   sphinx.ext.mathjax (7.2.6)
#   alabaster (0.7.16)
#   sphinxcontrib.applehelp (1.0.8)
#   sphinxcontrib.devhelp (1.0.6)
#   sphinxcontrib.htmlhelp (2.0.5)
#   sphinxcontrib.serializinghtml (1.1.10)
#   sphinxcontrib.qthelp (1.0.7)
#   matplotlib.sphinxext.plot_directive (3.10.0.dev151+gce15014066)
#   matplotlib.sphinxext.figmpl_directive (3.10.0.dev151+gce15014066)

# Traceback:
Traceback (most recent call last):
  File "/Users/aaronmeurer/miniconda3/envs/ndindex-matplotlib-dev/lib/python3.12/site-packages/sphinx/cmd/build.py", line 298, in build_main
    app.build(args.force_all, args.filenames)
  File "/Users/aaronmeurer/miniconda3/envs/ndindex-matplotlib-dev/lib/python3.12/site-packages/sphinx/application.py", line 355, in build
    self.builder.build_update()
  File "/Users/aaronmeurer/miniconda3/envs/ndindex-matplotlib-dev/lib/python3.12/site-packages/sphinx/builders/__init__.py", line 293, in build_update
    self.build(to_build,
  File "/Users/aaronmeurer/miniconda3/envs/ndindex-matplotlib-dev/lib/python3.12/site-packages/sphinx/builders/__init__.py", line 363, in build
    self.write(docnames, list(updated_docnames), method)
  File "/Users/aaronmeurer/miniconda3/envs/ndindex-matplotlib-dev/lib/python3.12/site-packages/sphinx/builders/__init__.py", line 571, in write
    self._write_serial(sorted(docnames))
  File "/Users/aaronmeurer/miniconda3/envs/ndindex-matplotlib-dev/lib/python3.12/site-packages/sphinx/builders/__init__.py", line 581, in _write_serial
    self.write_doc(docname, doctree)
  File "/Users/aaronmeurer/miniconda3/envs/ndindex-matplotlib-dev/lib/python3.12/site-packages/sphinx/builders/html/__init__.py", line 649, in write_doc
    self.docwriter.write(doctree, destination)
  File "/Users/aaronmeurer/miniconda3/envs/ndindex-matplotlib-dev/lib/python3.12/site-packages/docutils/writers/__init__.py", line 80, in write
    self.translate()
  File "/Users/aaronmeurer/miniconda3/envs/ndindex-matplotlib-dev/lib/python3.12/site-packages/sphinx/writers/html.py", line 36, in translate
    self.document.walkabout(visitor)
  File "/Users/aaronmeurer/miniconda3/envs/ndindex-matplotlib-dev/lib/python3.12/site-packages/docutils/nodes.py", line 186, in walkabout
    if child.walkabout(visitor):
       ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aaronmeurer/miniconda3/envs/ndindex-matplotlib-dev/lib/python3.12/site-packages/docutils/nodes.py", line 178, in walkabout
    visitor.dispatch_visit(self)
  File "/Users/aaronmeurer/miniconda3/envs/ndindex-matplotlib-dev/lib/python3.12/site-packages/sphinx/util/docutils.py", line 582, in dispatch_visit
    method(node)
  File "/Users/aaronmeurer/Documents/matplotlib/lib/matplotlib/sphinxext/figmpl_directive.py", line 174, in visit_figmpl_html
    imagedir, srcset, rel = _copy_images_figmpl(self, node)
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/aaronmeurer/Documents/matplotlib/lib/matplotlib/sphinxext/figmpl_directive.py", line 163, in _copy_images_figmpl
    shutil.copyfile(abspath, imagedir / name)
  File "/Users/aaronmeurer/miniconda3/envs/ndindex-matplotlib-dev/lib/python3.12/shutil.py", line 260, in copyfile
    with open(src, 'rb') as fsrc:
         ^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/wc/dppcpmxs1tlb36nqcw853wkm0000gn/T/pytest-of-aaronmeurer/pytest-2/test_srcset_version0/plot_directive/included_plot_21-1.2x.png'

```

asmeurer commented:
I added a test that will hopefully work. I'm having difficulty getting some of the tests to run successfully locally right now, but I believe the asserts I added work. If CI fails I might need some help figuring out why things aren't working on my computer.

tacaswell commented on lib/matplotlib/sphinxext/plot_directive.py:
```diff
@@ -775,14 +791,20 @@ def run(arguments, content, options, state_machine, state, lineno):
         base, ext = os.path.splitext(os.path.basename(source_file_name))
-        output_base = '%s-%d.py' % (base, counter)
+        if options['output-base-name']:
+            output_base = options['output-base-name'].format(counter=counter)
```
could we prefix in the script/rst file name here as well?  If you have duplicate output names in the same rst file that is a quick search to find and fix, but if you have the collision across multiple rst files it could be much more annoying to find where they are colliding.

asmeurer replied:
Well my thinking here is also that users should have full control over the filename that is produced. If we always inject something into the name, that won't be the case. For instance, if you right-click and save a plot, this will be the filename used for the image. 

Really, we need to figure out how to error if the same name is used twice. My naive idea of checking if the file already exists doesn't work because it could just exist from a previous build. It's probably possible to keep a global list of already used filenames, but I'll have to figure out how to make that work with partial Sphinx rebuilds.

tacaswell commented:
What version of pytest are you using?  I could not reproduce this issue last night.

tacaswell commented:
I am :+1: on this idea, one small question about the naming (I see both sides of forcing an additional prefix or not).

I re-milestoned this for 3.10 as it is a new feature which seems too big to put in via a micro release and it is too late to get this is under the wire for 3.9.

asmeurer commented:
> What version of pytest are you using? I could not reproduce this issue last night.

It's possible my local build of matplotlib is broken somehow. pytest also does a really annoying thing when I run this test on my Mac where it switches to another space like it is opening a GUI window. I also had to run 

```
sudo mkdir -p /tmp/.X11-unix
sudo chmod 1777 /tmp/.X11-unix
```

or the tests would just error completely with some strange error (which fortunately ChatGPT was able to figure out).

I tried my best to follow your dev instructions, but some of the steps didn't work (e.g., the dev requirements.txt doesn't actually list any build requirements).

tacaswell commented:
The reason I suspect pytest is this path: `'/private/var/folders/wc/dppcpmxs1tlb36nqcw853wkm0000gn/T/pytest-of-aaronmeurer/pytest-2/test_srcset_version0/plot_directive/included_plot_21-1.2x.png'` 

What was the strange error?  I'm really not sure why you need X11 stuff for the tests of the sphinx extension to run...


----

We have an open PR to fix the dev install instructions.  The conda environment file works, most of the regular devs who don't use conda know what to do, and we explicitly install them on CI so it fell through the cracks.

asmeurer commented:
Sorry if it wasn't clear. I reproduced the Sphinx issue outside of pytest, using

```
cd lib/matplotlib/tests/tinypages
python -m sphinx -W -b html -d doctrees . _build/html/ -D plot_srcset=2x
```

The first output I showed is from pytest (which is running that command). The second is the log file from running the command directly, which shows the full traceback.

asmeurer commented:
Ah, so looking closer at https://matplotlib.org/devdocs/devel/development_setup.html#create-a-dedicated-environment, I see there is a secret "conda" button you have to click. I really don't like that Sphinx extension that hides content behind tabs...

tacaswell commented:
oh, I bet our tiny pages config needs the backend set to `Agg`...

asmeurer commented:
The extension supposedly does that https://github.com/matplotlib/matplotlib/blob/eb172732c4fb3eaa692a1a5c7c4f5a528cc598ff/lib/matplotlib/sphinxext/plot_directive.py#L192

But perhaps my private matplotlib config is messing with things.

asmeurer commented:
I have some work-in-progress code to do the duplicate filename checks. I'm trying to figure out why this `out_of_date` function is needed https://github.com/matplotlib/matplotlib/blob/eb172732c4fb3eaa692a1a5c7c4f5a528cc598ff/lib/matplotlib/sphinxext/plot_directive.py#L484-L490. Shouldn't Sphinx already handle that logic itself? It apparently is needed based on this issue https://github.com/matplotlib/matplotlib/issues/17860, but I'm not understanding why.

asmeurer commented:
By the way, I saw you approved this. If you want to merge this as-is, that's fine. I can add the duplicate filename check in a separate PR. I'm still not completely sure yet if I can actually get it working.

asmeurer commented:
So it occurs to me that this might actually do unexpected things if a global output_base_name is set when there are partial rebuilds. I think I might just remove the ability to have a global option, as well as the `counter`. We can at least check that no names are the same document.

tacaswell commented:
This is also going to do funny looking things if you mix plots with and without set base names as the counter is always increased so I think you will get names like [`doc-1.png`, `my_name.png`, `doc-3.png`].

One option would be to the `'{counter}'` logic and move the counter incrementing into the other branch.

Another option would be to instead of thinking of this as setting the base name, change the user facing knob to be a postfix so either we append a counter _OR_ we stick on what ever the user gave us.  I think that will fix the myst problem and prevents inter-document name collisions.

asmeurer commented:
My plan is to simplify this considerably:

- Remove the global base-name flag, and also the `{counter}` thing. Just have `:output-base-name:` and that's it. Is it possible for a single `plot` directive to produce multiple files? If so, I'll still need to automatically append a counter in those cases. I think the issue you pointed out is accurate so I'll need to update the counter to work per-base name. If that's not possible, we can just ignore the counter completely when base-name is set. 

- Add a basic check for duplicate names in the same file. Checking for duplicate names across files is doable but annoying because we have to store information in the Sphinx environment so that it works across rebuilds.

tacaswell commented:
You can have multiple figures in one `.. plot` (see https://github.com/matplotlib/matplotlib/blob/955432c552d6da7250a76c4f623b33b5d4e6c215/lib/matplotlib/tests/tinypages/some_plots.rst?plain=1#L161-L168)

ksunden commented:
Bumped to 3.11 as we are preparing 3.10 and the latest comment here is that changes will be made.

asmeurer commented:
I've updated this in the way that I suggested above:

- I removed the `{counter}` thing. There's now just `:output-base-name:`. Plots named in this way do not use a counter (excepting as noted below).
- Added a global check that no two `.. plot` directives use the same `output-base-name`. Testing this would require adding a completely separate test docs build. This seemed like a lot for one test, but I can do if you want. I did test it manually. 
- I also removed the public `plot_output_base_name`, as it no longer makes sense.
- I still have the test failure above locally. I haven't yet figured out how to work around that (as I noted, I suspect my local tests aren't setting the plotting backend correctly for some reason).

> You can have multiple figures in one .. plot (see

It turns out this actually uses a separate counter (that example generates `some_plots-16_00.png` and `some_plots-16_01.png`). This also applies to plot names generated with `:output-base-name:` (it's done in a completely separate place in the code. This could probably be cleaned up, but I don't particularly feel like doing it).

timhoffm commented on lib/matplotlib/sphinxext/plot_directive.py:
```diff
@@ -775,14 +830,22 @@ def run(arguments, content, options, state_machine, state, lineno):
         code = textwrap.dedent("\n".join(map(str, content)))
-        counter = document.attributes.get('_plot_counter', 0) + 1
-        document.attributes['_plot_counter'] = counter
         base, ext = os.path.splitext(os.path.basename(source_file_name))
```
Small nit, but if I see it correctly, this is only used in the exiles branch, so could be moved there.

timhoffm commented on lib/matplotlib/sphinxext/plot_directive.py:
```diff
@@ -775,14 +830,22 @@ def run(arguments, content, options, state_machine, state, lineno):
         base, ext = os.path.splitext(os.path.basename(source_file_name))
-        output_base = '%s-%d.py' % (base, counter)
+        if options['output-base-name']:
+            output_base = options['output-base-name']
```
Don't we need a .py extension here like in the else case?

asmeurer replied:
This block handles the case where there isn't an extension. https://github.com/matplotlib/matplotlib/pull/28187/files#diff-1565d78306cf72de5a28f4e2b196ee2cb09026d1f8aa2dc26178c3a47f3347b8R853-R856

timhoffm commented on lib/matplotlib/sphinxext/plot_directive.py:
```diff
@@ -47,6 +47,12 @@
 
 The ``.. plot::`` directive supports the following options:
 
+``:output-base-name:`` : str
```
Is this really only the filename base or can it be a relative or absolute path without extension? Either case should be documented. Also, since you start giving users control over the created files, you probably should mention where they go.

asmeurer replied:
It's just the file name. The files go where ever Sphinx puts them (it looks like it's in the `_images` directory, but I don't know if that's configurable).

timhoffm replied:
Thanks for the clarification. In that case `*-base-name` makes sense.

Two further thoughts:
- would `image-base-name` be better (because more concrete) than `output-base-name`?
- what happens / do we need to check and error if the user tries to give more than a base name, e.g. `:output-base-name: ../escaped_from_image_dir/myimage` or `nested_dir/myimage`?

timhoffm commented on lib/matplotlib/tests/tinypages/some_plots.rst:
```diff
@@ -174,3 +174,18 @@ Plot 21 is generated via an include directive:
 
 .. plot:: range6.py range10
+
+Plots 23 through 25 use output-base-name.
```
Why do we need 3? / What do we get more compared to one plote?

asmeurer replied:
The second one probably isn't needed (it was when I had the counter thing but I removed that). The third one is testing the case when the plot comes from a file, which is a completely separate code path in the extension.

asmeurer commented:
> what happens / do we need to check and error if the user tries to give more than a base name, e.g. :output-base-name: ../escaped_from_image_dir/myimage or nested_dir/myimage

Good question. I'll need to test it. Actually a more common error would be trying to add the extension to the base name, like `myimage.png`. Maybe we should disallow that?

timhoffm commented:
Yes, extensions should not be allowed. Possibly slightly more restrictive than need be, but for simplicity I'd go with not allowing `.` and `/` in the name. People are welcome to propose which corner case is supported by a more refined check.

timhoffm commented:
I'd go with `image-basename`. Pulling `basename` together follows the notation of `os.path.basename` and also base and name are stronger bound together than image.
Using `image` instead of `output` feels more natural. If you ask, "what does the command produce write to disk?" The answer is more likely "an image file" rather than an "output file". "Image" is more concrete. "Output" would only be preferable if the image could also be interpreted as an input, but that cannot happen in the semantic context of the directive.

asmeurer commented:
OK, I've renamed `output-base-name` to `image-basename`.

timhoffm commented on lib/matplotlib/sphinxext/plot_directive.py:
```diff
@@ -312,9 +321,37 @@ def setup(app):
+            del env.mpl_custom_base_names[docname]
+
+    def merge_other(self, app, env, docnames, other):
+        for docname in docnames:
```
Do we need to loop over the docnames? AFAICS, we only want to merge `other.mpl_custom_base_names` into `env.mpl_custom_base_names`. I.e. we can directly loop over that.

asmeurer replied:
OK, I simplified this function. I'm not completely sure about it, but the functionality still seems to work based on manual testing.

timhoffm commented on lib/matplotlib/sphinxext/plot_directive.py:
```diff
@@ -312,9 +321,37 @@ def setup(app):
+        for docname in docnames:
+            if docname in other.mpl_custom_base_names:
+                if docname not in env.mpl_custom_base_names:
+                    env.mpl_custom_base_names[docname] = set()
```
Can be left out because mpl_custom_base_names is a defaultdict.

timhoffm commented on lib/matplotlib/sphinxext/plot_directive.py:
```diff
@@ -312,9 +321,37 @@ def setup(app):
+def init_filename_registry(app):
+    env = app.builder.env
+    if not hasattr(env, 'mpl_custom_base_names'):
+        env.mpl_custom_base_names = defaultdict(set)
```
The name is a bit generic. Maybe 
```suggestion
        env.mpl_plot_image_basenames = defaultdict(set)
```
Or similar. This naming is *library_directive_option*.

asmeurer commented:
I think the docs build error isn't my fault this time, from what I can tell.

asmeurer commented:
Any more thoughts on this?

timhoffm commented:
We just need a second core developer to approve.

asmeurer commented:
@tacaswell you looked at this previously. Do you have any thoughts?

asmeurer commented:
Can this be merged? It's been sitting for some time without any further review.

timhoffm commented:
We still need a second review from a core developer.

jklymak commented:
- circleCI is failing, and the log is now gone, so it's pretty hard to approve a doc PR if the docs don't build cleanly.  Our doc tests currently are working, so please rebase.
- Has anyone tested this for performance?  This has a loop through all the basenames - is that a trivial call for every image in a gallery?  I guess we don't have _that_ many plot directives, so maybe it is OK.

asmeurer commented:
> Has anyone tested this for performance? This has a loop through all the basenames - is that a trivial call for every image in a gallery? I guess we don't have that many plot directives, so maybe it is OK.

If you're talking about `_FilenameCollector`, all that's doing is creating a mapping of plots to filenames, so that Sphinx can persist it between runs.

timhoffm commented:
No it's a flaky test.

QuLogic commented on lib/matplotlib/sphinxext/plot_directive.py:
```diff
@@ -600,6 +635,27 @@ def _parse_srcset(entries):
+        if output_base in env.mpl_plot_image_basenames[d]:
+            if d == docname:
+                raise PlotError(
+                    f"The image-basename "
+                    f"{output_base}' is used multiple times.")
```
Fits in a line; also the quotes are unbalanced:
```suggestion
                    f"The image-basename {output_base!r} is used multiple times.")
```

QuLogic commented on lib/matplotlib/sphinxext/plot_directive.py:
```diff
@@ -600,6 +635,27 @@ def _parse_srcset(entries):
+                raise PlotError(
+                    f"The image-basename "
+                    f"{output_base}' is used multiple times.")
+            raise PlotError(f"The image-basename "
+                            f"'{output_base}' is used multiple times "
+                            f"(it is also used in {env.doc2path(d)}).")
```
```suggestion
            raise PlotError(f"The image-basename {output_base!r} is used multiple"
                            f"times (it is also used in {env.doc2path(d)}).")
```

QuLogic commented on lib/matplotlib/sphinxext/plot_directive.py:
```diff
@@ -723,6 +779,7 @@ def render_figures(code, code_path, output_dir, output_base, context,
 def run(arguments, content, options, state_machine, state, lineno):
     document = state_machine.document
     config = document.settings.env.config
+    env = document.settings.env
```
```suggestion
    env = document.settings.env
    config = env.config
```

QuLogic reviewed (commented):
Something missing from the tests I think is what happens when `image-basename` is set and the `plot` directive creates multiple figures? I don't believe this is currently handled at least based on the diff of `_plot_counter`.

QuLogic commented on lib/matplotlib/sphinxext/plot_directive.py:
```diff
@@ -600,6 +635,27 @@ def _parse_srcset(entries):
+def check_output_base_name(env, output_base):
+    docname = env.docname
+
+    if '.' in output_base or '/' in output_base:
```
Since this is passed through `Path`, etc., you should also check for `\\`.

QuLogic commented:
I can't comment on the line directly since it's not part of the diff, but the `output_base` variable that was changed is also used for `show-source-link`, which doesn't look documented or tested (and thus I'm not sure if that was intentional either.)

asmeurer commented:
I added a test for multiple figures. I can confirm that it does work by testing manually, but for some reason adding leads to a much earlier test failing. 

```

tmp_path = PosixPath('/private/var/folders/wc/dppcpmxs1tlb36nqcw853wkm0000gn/T/pytest-of-aaronmeurer/pytest-6/test_tinypages0')

    def test_tinypages(tmp_path):
        shutil.copytree(tinypages, tmp_path, dirs_exist_ok=True)
        html_dir = tmp_path / '_build' / 'html'
        img_dir = html_dir / '_images'
        doctree_dir = tmp_path / 'doctrees'

        # Build the pages with warnings turned into errors
        build_sphinx_html(tmp_path, doctree_dir, html_dir)

        def plot_file(num):
            return img_dir / f'some_plots-{num}.png'

        def plot_directive_file(num):
            # This is always next to the doctree dir.
            return doctree_dir.parent / 'plot_directive' / f'some_plots-{num}.png'

        range_10, range_6, range_4 = (plot_file(i) for i in range(1, 4))
        # Plot 5 is range(6) plot
        assert filecmp.cmp(range_6, plot_file(5))
        # Plot 7 is range(4) plot
        assert filecmp.cmp(range_4, plot_file(7))
        # Plot 11 is range(10) plot
        assert filecmp.cmp(range_10, plot_file(11))
        # Plot 12 uses the old range(10) figure and the new range(6) figure
        assert filecmp.cmp(range_10, plot_file('12_00'))
        assert filecmp.cmp(range_6, plot_file('12_01'))
        # Plot 13 shows close-figs in action
        assert filecmp.cmp(range_4, plot_file(13))
        # Plot 14 has included source
        html_contents = (html_dir / 'some_plots.html').read_text(encoding='utf-8')

        assert '# Only a comment' in html_contents
        # check plot defined in external file.
>       assert filecmp.cmp(range_4, img_dir / 'range4.png')
E       AssertionError: assert False
E        +  where False = <function cmp at 0x12710b560>(PosixPath('/private/var/folders/wc/dppcpmxs1tlb36nqcw853wkm0000gn/T/pytest-of-aaronmeurer/pytest-6/test_tinypages0/_build/html/_images/some_plots-3.png'), (PosixPath('/private/var/folders/wc/dppcpmxs1tlb36nqcw853wkm0000gn/T/pytest-of-aaronmeurer/pytest-6/test_tinypages0/_build/html/_images') / 'range4.png'))
E        +    where <function cmp at 0x12710b560> = filecmp.cmp

lib/matplotlib/tests/test_sphinxext.py:75: AssertionError
```

asmeurer commented:
Good catch on the source link. It is working, but it looks like it isn't properly adding the `.py` extension when there is a custom name. I'm working on a fix.

asmeurer commented:
Oh I think that was just because I was testing tinypages locally and it was messing with the tests.

asmeurer commented:
OK, this is ready for review again.

QuLogic commented on lib/matplotlib/tests/test_sphinxext.py:
```diff
@@ -16,6 +16,17 @@
 tinypages = Path(__file__).parent / 'data/tinypages'
 
 
+def _ignore_build_artifacts(dir, files):
```
From the docs, you can use `shutil.ignore_patterns` to generate this function directly.

QuLogic reviewed (commented):
Since this also affects the naming of the source file, should we go back to `output-basename`, or are you comfortable with it staying `image-basename`, @timhoffm?

timhoffm commented:
> Since this also affects the naming of the source file, should we go back to `output-basename`, or are you comfortable with it staying `image-basename`, @timhoffm?

I don't care too much between `output` and `image`. One can argue both ways: The source and the image use this, or we focus on the eventually desired output (the image) and the source just follows along with it.

We should consider whether *basename* is appropriate. While [os.path.basename](https://docs.python.org/3/library/os.path.html#os.path.basename) includes the extension, here it's not included. [stem](https://docs.python.org/3/library/pathlib.html#pathlib.PurePath.stem) would be the technically correct term. However I feel `output-stem` or `image-stem` is not instructive; we'd need to go with the lengthy  `output-filename-stem` or `image-filename-stem`. Maybe `*-prefix` would be another alternative.

Either way, I believe the logic is too complex to caputure everything in a single name. Until now, everything was magically determined behind the scenes (i.e. users should not have to bother with the exact name). If we now make the name configurable, we should explain what exactly this is in the associated docs

(1) this must not have any path components
(2) it may be the full filename stem, or only a prefix since AFAICS there are multiple ways that additional components could be appended.

asmeurer commented:
My original version did have something more configurable (the user could specify the format string), but I decided that was way too complex for what was needed here. I think the main thing here is that the user has some way to make the name a little more human readable. If it also adds some numbering or file extensions to that it isn't a problem, especially given that no one else has even raised this as a problem before. And obviously for my original problem the name itself doesn't even matter; it just needs to be unique.

timhoffm commented:
Then possibly `output-prefix` or `image-prefix` are good names as they only claim to define a part of the filename.

asmeurer commented:
FWIW, none of the suggested names really make a ton of sense if you just see them in a `plot` directive without the context of the sphinx extension internals, or even just the context of the matplotlib sphinx extension documentation. Perhaps the name should just be something like `output-name` or `filename-prefix` that is a little more self-documenting.

asmeurer commented:
OK, I've updated it to use `filename-prefix`.

timhoffm commented:
Failing tests should be fixed by #30162. @asmeurer can you please rebase on main?
